src/tcpip/tcpip.c

```diff
@@ -823,20 +787,20 @@ static void mg_tcpip_poll(struct mg_tcpip_if *ifp, uint64_t uptime_ms) {
 // somewhere fast. Note that newlib's malloc is not thread safe, thus use
 // our lock-free queue with preallocated buffer to copy data and return asap
 void mg_tcpip_qwrite(void *buf, size_t len, struct mg_tcpip_if *ifp) {
-  if (q_write(&ifp->queue, buf, len)) {
+  char *p;
+  if (mg_queue_book(&ifp->recv_queue, &p, len) >= len) {
+    memcpy(p, buf, len);
+    mg_queue_add(&ifp->recv_queue, len);
     ifp->nrecv++;
   } else {
     ifp->ndrop++;
+    MG_ERROR(("Recv queue full: %lu/%lu", len, ifp->recv_queue.size));
   }
 }
 
-size_t mg_tcpip_qread(void *buf, struct mg_tcpip_if *ifp) {
-  return q_read(&ifp->queue, buf);
-}
-
 size_t mg_tcpip_driver_rx(void *buf, size_t len, struct mg_tcpip_if *ifp) {
-  return mg_tcpip_qread((void *) ifp->rx.ptr, ifp);
-  (void) len, (void) buf;
+  (void) buf, (void) len, (void) ifp;
+  return 0;
 }
```

Comment:
This allows drivers to just put a pointer to this function in the function structure, a generic poll from queue function that works for all IRQ based drivers, and avoid implementing that themselves.

Reply:
OK, this one will go as soon as we fix the RMII example...